### ISSUE TITLE
How to test my own images from a video clip?

### ISSUE THREAD

**Issue by fjremnav** [jfzhuang/DAVSS#2]
I like to test your pretrained cityscapes model with my own video clip.
I can extract each frame from my own video, but I don't see you have a script to do inference of test images from a video.
Can you provide a guidance of how to do this?

Thanks,

**Comment by jfzhuang** [comment]
Actually, the show.sh is used to generate segmentation results. You can modify the script to meet your requirement.

**Comment by fjremnav** [comment]
@jfzhuang 

  I look at show.sh and show.py and  the code requires ground truth. But my own images do not have ground truths.

Thanks,

**Comment by jfzhuang** [comment]
You need to make the following modifications, such as

1. modify the dataset class so that it don't need gt path and return only image sequence
2. modify the show.py to only decode segmentation results rather than ground truth

**Comment by fjremnav** [comment]
I am able to modify your code to make it work without groundtruth.
However, I am not able to see much of temporal consistency improvement compared with baseline.
Maybe I didn't modify the code properly because I just want to improve temporal consistency.
I modify show.py (dataloader is modified too) as below:
for t in range(1:N):
                    img_t-1=img_t-1.cuda()
                    feat = deeplab(img_t-1)
                    feat_up_1 = F.interpolate(feat, scale_factor=4, mode='bilinear', align_corners=True)
                    feat_rgb = pred2im(feat_up_1)
                    cv2.imwrite(os.path.join(args.save_path, 'base_{}.png'.format(t-1)), feat_rgb) #(baseline)
                    img_t =img_t.cuda()
                    flow = flownet(torch.cat([img_t, img_t-1], dim=1))
                    feat = warpnet(feat, flow)
                    warp_im = F.upsample(img_1, scale_factor=0.25, mode='bilinear', align_corners=True)
                    warp_im = warpnet(warp_im, flow)
                    img_t_down = F.upsample(img_t, scale_factor=0.25, mode='bilinear', align_corners=True)
                    dm = dmnet(warp_im, img_2_down)
                    dm_up = F.interpolate(dm, scale_factor=4, mode='bilinear', align_corners=True)

                    feat_cc = cfnet(img_t)
                    feat_cc_up = F.interpolate(feat_cc, scale_factor=4, mode='bilinear', align_corners=True)

                    feat_up = F.interpolate(feat, scale_factor=4, mode='bilinear', align_corners=True)
                    feat_merge = feat_up * (1-dm_up) + feat_cc_up*dm_up

                    feat_merge_rgb = pred2im(feat_merge)
                    cv2.imwrite(os.path.join(args.save_path, '{}.png'.format(t)), feat_merge_rgb)

I just use t-1 img to correct t img segmentation to improve temporal consistency. Is the above change correct?

Thanks,

**Comment by jfzhuang** [comment]
Your code seems correct. Actually, our method is proposed to improve segmentation accuracy rather than temporal consistency. Maybe you can survey other video segmentation methods. Welcome to further discussion if you have any ideas.

**Comment by fjremnav** [comment]
I am able to improve temporal consistency with your pretrained weights.
However, it seems segmentation quality of baseline deeplabv3plus is not so good and hence it impacts to DAVSS quality.
Is there a way to improve deeplabv3plus quality? Is it possible to use  an existing model, like the one below which can achieve much better quality than the deeplabv3plus in the DAVSS?
https://github.com/NVIDIA/semantic-segmentation.git

My idea is to use semantic-segmentation pretrained weight as it is (no need to retrain) and reuse the rest of DAVSS components to achieve better performance

Thanks,

**Comment by jfzhuang** [comment]
You can try some other advanced segmentation models, such as HRNet and OCRNet. Waiting for your experimental results.

**Comment by fjremnav** [comment]
I don't have a GPU server powerful enough to train semantic segmentation model and it will be better to use existing pretrained model .Do you have a guidance of how to do this, ie, no retraining on baseline model into your DAVSS pipeline?

Thanks,

**Comment by jfzhuang** [comment]
For example, mmsegmentation project provides many advanced segmentation models with pretrained weights, which can be used to replace the segmentation model in DAVSS. However, you need to retrain the DAVSS, since other components, like dmnet, flownet and cfnet, need to train with the segmentation model for compatibility.

**Comment by fjremnav** [comment]
Is it possible to use pretrained deeplabv3plus from mmsegmentation instead of using your pretrained model?
I suspect your trained deeplabv3plus performance may not be as good as one from mmsegmentation.
Have you compared this?

Thanks,

**Comment by jfzhuang** [comment]
Yes, my trained deeplabv3plus achieves 76.61% mIoU score on Cityscapes while the mmsegmentation one achieves near 80%. You can use the mmsegmentation one for experiments.

**Comment by fjremnav** [comment]
How do I use pretrained deeplabv3plus model of mmsegmentation in your show.py?
Thanks,

**Comment by jfzhuang** [comment]
You need to replace the segmentation model in scnet.py and retrain other components, including dmnet, flownet and cfnet. After that, you can use show.py by modifying the checkpoint path.

**Comment by fjremnav** [comment]
Can you explain how feat is done in warpnet below:
feat = warpnet(feat, flow)

since feat shape is (1, 19, 480, 270), do you apply warpnet to feat[0,i,:,:] for i in range (19)?

Thanks,

**Comment by fjremnav** [comment]
One more question, how can I do flownet at native resolution, 1920x1080?
Your current code, flownet and warpnet work at 480x270. I like to work on a native resolution, 1920x1080.

FC

**Comment by jfzhuang** [comment]
Actually, feature warping operation is first used in DFF (Deep Feature Flow). You can refer to the paper and code of DFF for further understanding.
